### PR TITLE
Refactors into cursor class

### PR DESCRIPTION
**Pull request by myxoh**
Breakup from: https://github.com/xing/rails_cursor_pagination/pull/68

Refactors into `Cursor` class

The purpose of these changes were us adapting our in-house pagination to this library without having to modify our FE client; as such we had to make three modifications for the following reasons.

**Refactoring into the Cursor class**
So this one was again an important one for us. Our pagination right now works using a UUID column, regardless of which ordering column the FE requires.
We felt it redundant to introduce yet another attribute for our FE to handle, when the UUID already refers to a unique tuple of id / order field (by virtue of being part of a record).
As such - the compromise solution we came up with was - by extracting the encoding / decoding to a helper class; we were able to create our cursors on the BE based on the FE information given - rather than require the FE to send the cursor (which would require us exposing it, and the FE consuming and sending it back).

### CODE REVIEW

**Review by nicolas-fricke — COMMENTED** [review]
I really like the effort of extracting the cursor into its own class 🎉 Thanks a lot for that! I'd say we should definitely add some good tests coverage for the class itself before merging it.

However, while reviewing I thought about the use-case you're describing there: to expose a different kind of cursor to the client. But the current approach still means that the standard cursors will be sent in the `Paginator#fetch` pagination response within the `page_info` and under each `page[i][:cursor]` field, right?

So to be fully able to paginate using different cursors, we should then provide a way of overwriting how the cursor that's sent to the client looks, right? Apart from monkey-patching the `Cursor#encode` method (what I'd strongly advise against as it's prone to break), we could consider allowing to configure an alternative `Cursor` class that follows the same interface.

Ultimately, if we document this as an idea to use different cursors, all cursors that are visible to the client should match that alternative format (be it UUIDs or whatever else).

What do you think?

**Comment by myxoh** [comment]
> I really like the effort of extracting the cursor into its own class 🎉 Thanks a lot for that! I'd say we should definitely add some good tests coverage for the class itself before merging it.
> 
> However, while reviewing I thought about the use-case you're describing there: to expose a different kind of cursor to the client. But the current approach still means that the standard cursors will be sent in the `Paginator#fetch` pagination response within the `page_info` and under each `page[i][:cursor]` field, right?
> 
> So to be fully able to paginate using different cursors, we should then provide a way of overwriting how the cursor that's sent to the client looks, right? Apart from monkey-patching the `Cursor#encode` method (what I'd strongly advise against as it's prone to break), we could consider allowing to configure an alternative `Cursor` class that follows the same interface.
> 
> Ultimately, if we document this as an idea to use different cursors, all cursors that are visible to the client should match that alternative format (be it UUIDs or whatever else).
> 
> What do you think?

I'll address the actual comments later in the week but thought I should respond to this first.

In our case - after our migration to the gem we are ignoring the `cursor` in the response and just using the records. I think there was a suggestion elsewhere on an issue to expose just the records.

In fact, the main point of bypassing the `cursor` on this library in our case is we don't want to send unnecessary extra data to the FE (also  we don't want to expose the internal ids which we would indirectly do with the current cursor response)

I feel it'd probably be a bit heavy-handed to attempt to fit any custom cursor into the library and respond using those values. And I'd probably stick with either using the default cursor from the library, and the library as-is; or if you want to go custom just pulling the records (which I was planning to add an MR for) and using a custom cursor logic.

In summary - I'd probably avoid trying to anticipate how a consumer might want to use the cursors - other than providing a sensible default (that the library already does); and rather just expose a few tools for those who want to opt-out of those defaults

File: lib/rails_cursor_pagination/cursor.rb [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      raise InvalidCursorError,
+            "The given cursor `#{@cursor.inspect}` could not be decoded"
+    end
+
+    # Initializes the record
+    #
+    # @param id [Integer] the ID of the cursor record.
+    # @param order_field [Symbol] the column or virtual column for ordering.
+    # @param field the value that the order_field column on the cursor record
+    # (if the order field is not an id).
```
[the remark this answers is not in the document]
Comment (myxoh) [reply lib/rails_cursor_pagination/cursor.rb]: Yep, sorry for that!

File: lib/rails_cursor_pagination/cursor.rb [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    end
+
+    # Decodes an encoded cursor
+    # Decode the provided cursor. Either just returns the cursor's ID or in case
+    # of pagination on any other field, returns a tuple of first the cursor
+    # record's other field's field followed by its ID.
+    #
+    # @param encoded [String] encoded cursor
+    # @return [Integer, Array]
+    def self.decode(encoded, order_field = :id)
```
[the remark this answers is not in the document]
Comment (myxoh) [reply lib/rails_cursor_pagination/cursor.rb]: Yes, I agree it reads better and it's more accurate this way.

**Comment by myxoh** [comment]
OK - I've added a number of tests covering all the public interface of the new class.

The failing tests seems to be down to a connection failure in MySQL so I think it's safe to re-run.

Let me know your thoughts

**Review by nicolas-fricke — COMMENTED** [review]
Thanks for the refactoring and for sharing your thoughts. I agree that trying to anticipate how others might use the gem is not ideal. So if there's no need to support custom cursors classes, we can probably live without it.

However, I think it's then maybe not ideal to document the idea of manually constructing the cursor the way we now have it in this PR, as that would lead to a weird in-between state where users can use different cursors, but will still get the "wrong" ones from the gem. So how about in this PR only extracting the cursor to its own class (something that was on our list before), but removing the two comments on how to use the class outside of the gem.

If then someone wants to play around with it on their own, it's up to them. But then there's maybe less of an expectation that the gem supports custom cursors as of now.

Would that be a valid path forward for you?

**Comment by myxoh** [comment]
> Thanks a lot for the refactoring and for sharing your thoughts. I agree that trying to anticipate how others might use the gem is not ideal. So if there's no need to support custom cursors classes, we can probably live without it.
> 
> However, I think it's then maybe not ideal to document the idea of manually constructing the cursor the way we now have it in this PR, as that would lead to a weird in-between state where users can use different cursors, but will still get the "wrong" ones from the gem. So how about in this PR only extracting the cursor to its own class (something that was on our list before), but removing the two comments on how to use the class outside of the gem.
> 
> If then someone wants to play around with it on their own, it's up to them. But then there's maybe less of an expectation that the gem supports custom cursors as of now.
> 
> Would that be a valid path forward for you?

Yes - I understand your hesitation; specially in the current shape it feels like a very hacky path to be documented. We can keep this PR  as a pure refactor and we'll use the internals on our project at our own risk which I feel is a reasonable trade-off

File: spec/rails_cursor_pagination/cursor_spec.rb [hunk 3]
```diff
@@ -0,0 +1,203 @@
+# frozen_string_literal: true
+
+RSpec.describe RailsCursorPagination::Cursor do
+  describe '.encode' do
+    let(:record) { Post.create! id: 1, author: 'John', content: 'Post 1' }
+
+    context 'when ordering by id implicitly' do
+      subject(:encoded) do
+        described_class.encode(record: record)
+      end
```
[the remark this answers is not in the document]
Comment (nicolas-fricke) [reply spec/rails_cursor_pagination/cursor_spec.rb:10]: Thinking about it more, I actually believe given the alternatives, the easiest and cleanest way would be to remove the class-method `.encode` and just call `Cursor.new(...).encode` instead of the `Cursor.encode(...)` shortcut. 

I believe that all methods that are accessible from the outside should have proper test coverage – even more so in a gem that might be used (and abused or monkey-patched 🙈) by others. Then it's important to make sure that the methods we offer to the outside do exactly what we want them to and that this behavior stays consistent between versions (unless it's a conscious decision to change it, of course). But therefore, it also makes sense to keep the public API as small as possible.

And I would refrain from using something like `.send` meta-programming just to reduce the API of the class. That makes understanding the code especially for beginners (as well as some more naïve debuggers or IDEs) unnecessarily hard.

But I actually also see explicitly calling `Cursor.new(...).encode` as a benefit: it shows that there's an actual cursor object that's being created and used – something that might help others in future if they should want to implement their own cursor classes or use the Cursor from the outside.

So **tl;dr**: My suggestion would be to get rid of `Cursor.encode` in favor of just using (and testing) the instance method `Cursor#encode` 🙂 

What do you think??

**Review by nicolas-fricke — COMMENTED** [review]
Amazing! Just one last tiny comment regarding the `describe` and then I believe we're ready to merge this 🎉

File: spec/rails_cursor_pagination/cursor_spec.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (nicolas-fricke) [reply spec/rails_cursor_pagination/cursor_spec.rb]: Looks great, thanks a lot for the changes!! 😊

**Review by nicolas-fricke — APPROVED** [review]
Awesome, thanks so much for extracting this into its own class! 👏
